tools/build/third_party/breakpad/dump_syms_pe.cc

```diff
@@ -0,0 +1,389 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Windows utility to dump the line number data from a .exe built
+// with debug info with MinGW's gcc to the minidump text-based
+// format for breakpad.
+
+#include <cstring>
+#include <iostream>
+#include <fstream>
+#include <stdio.h>
+#include <string>
+
+#include "common/dwarf/bytereader-inl.h"
+#include "common/dwarf/dwarf2diehandler.h"
+#include "common/dwarf_cfi_to_module.h"
+#include "common/dwarf_cu_to_module.h"
+#include "common/dwarf_line_to_module.h"
+#include "common/linux/dump_symbols.h"
+#include "common/module.h"
+#include "common/path_helper.h"
+
+
+typedef struct PEFileHeader {
+  uint16_t machine;
+  uint16_t numSections;
+  uint32_t timeStamp;
+  uint32_t symbolTableOffset;
+  uint32_t numSymbols;
+  uint16_t extraHeaderSize;
+  uint16_t flags;
+} PEFileHeader;
+
+typedef struct PEOptionalHeader {
+  uint16_t signature;
+  uint16_t linkerVersion;
+  uint32_t sizeOfCode;
+  uint32_t sizeOfInitData;
+  uint32_t sizeOfUnintData;
+  uint32_t entryPoint;
+  uint32_t codeBase;
+  uint64_t imageBase;
+  uint32_t sectionAlign;
+  uint32_t fileAlign;
+  uint32_t osVersion;
+  uint32_t imageVersion;
+  uint32_t subsystemVersion;
+  uint32_t winVersion;
+  uint32_t imageSize;
+  uint32_t headerSize;
+  uint32_t checksum;
+  uint16_t subsystem;
+  uint16_t dllFlags;
+  uint64_t stackReserve;
+  uint64_t stackCommit;
+  uint64_t heapReserve;
+  uint64_t heapCommit;
+  uint32_t loaderFlags;
+  uint32_t numDirEntries;
+  struct dirEntry {
+    uint32_t address;
+    uint32_t size;
+  } directory[7]; // We only care about the debug entry, which is the 7th.
+} PEOptionalHeader;
+
+typedef struct PESectionHeader {
+  char name[8];
+  uint32_t virtualSize;
+  uint32_t virtualAddress;
+  uint32_t size;
+  uint32_t offset;
+  uint32_t reloctionsOffset;
+  uint32_t lineNumbersOffset;
+  uint16_t numRelocations;
+  uint16_t numLineNumbers;
+  uint32_t flags;
+} PESectionHeader;
+
+#define SYMBOL_SIZE 18
+
+typedef struct StringTable {
+  const char* data;
+  int size;
+} StringTable;
+
+typedef struct PEFile {
+  std::vector<char> data;
+  PEFileHeader* header;
+  PEOptionalHeader* optionalHeader;
+  PESectionHeader* sections;
+  StringTable strings;
+  std::string debugId;
+  std::string codeId;
+} PEFile;
+
+typedef struct DebugDirectory {
+  uint32_t signature;
+  uint32_t timestamp;
+  uint32_t version;
+  uint32_t type;
+  uint32_t size;
+  uint32_t address;
+  uint32_t offset;
+} DebugDirectory;
+
+typedef struct DebugData {
+  uint32_t signature;
+  struct {
+    uint32_t a;
+    uint16_t b;
+    uint16_t c;
+    uint8_t d[8];
+  } uuid;
+  uint32_t age;
+} DebugData;
+
+static int getBase(const std::vector<char>& data) {
```

Comment:
Modern C++ suggests putting these in an anonymous namespace instead of making them all static. (Then you don't have to mark them all as static)